File: analysis/65_audit_v2_metadata_coverage.R

options(stringsAsFactors = FALSE)

# Sepsis V2 - Step 65 metadata coverage audit
# Designed for R 4.4.0 on Windows

pkgs <- c("readr","readxl","dplyr","stringr","purrr","tibble","xml2")
missing <- pkgs[!vapply(pkgs, requireNamespace, logical(1), quietly = TRUE)]
if (length(missing) > 0) {
  install.packages(missing, repos = "https://cloud.r-project.org")
}

suppressPackageStartupMessages({
  library(readr)
  library(readxl)
  library(dplyr)
  library(stringr)
  library(purrr)
  library(tibble)
  library(xml2)
})

DATA_ROOT <- "E:/sepsis_project/data"
OUT_ROOT  <- "E:/sepsis_project/results/V2_05_metadata_coverage"
dir.create(OUT_ROOT, recursive = TRUE, showWarnings = FALSE)

PROJECTS <- c(
  "PRJEB37289","PRJEB67798","PRJEB82425",
  "PRJNA516701","PRJNA578267","PRJNA595346","PRJNA884103"
)

patterns <- list(
  patient_id = c("\\bpatient\\b","\\bsubject\\b","participant","patient.?id","subject.?id"),
  sample_id = c("sample.?id","sample_accession","biosample","sequence.?name","sample.?name","specimen"),
  run_id = c("run_accession","sra.?run","\\brun.?id\\b","\\bsrr[0-9]+","\\berr[0-9]+","\\bcrr[0-9]+"),
  time = c("time.?point","timepoint","\\bday\\b","\\bdol\\b","collection.?date","sampling.?date","visit","baseline","discharge","intubation","icu.?day","hospital.?day"),
  infection_group = c("\\bsepsis\\b","septic.?shock","\\bvap\\b","pneumonia","infected","bacteremia","bacteraemia","bloodstream.?infection","\\bbsi\\b"),
  infection_source = c("infection.?source","infection.?site","site.?of.?infection","focus.?of.?infection","pulmonary.?infection","urinary.?infection","abdominal.?origin","bloodstream.?infection"),
  outcome = c("\\boutcome\\b","mortality","\\bdeath\\b","survival","survivor","deceased","icu.?length.?of.?stay","hospital.?length.?of.?stay"),
  antibiotics = c("antibiotic","antimicrobial","\\babx\\b","cefazolin","clindamycin","tobramycin","colistin","cefotaxime","antibiotic.?score"),
  severity = c("\\bsofa\\b","\\bapache\\b","severity","shock","vasopressor","mechanical.?ventilation","ventilation.?duration"),
  demographics = c("\\bage\\b","\\bsex\\b","gender","birth.?weight","gestational.?age","\\bbmi\\b"),
  body_site = c("body.?site","sample.?type","rectal","stool","faecal","fecal","tracheal","endotracheal","gut","lung")
)

detect_domains <- function(text) {
  text <- tolower(paste(text, collapse = " "))
  vapply(patterns, function(pats) {
    any(vapply(pats, function(p) str_detect(text, regex(p, ignore_case = TRUE)), logical(1)))
  }, logical(1))
}

read_tabular <- function(path) {
  ext <- tolower(tools::file_ext(path))
  if (ext == "csv") {
    x <- tryCatch(read_csv(path, n_max = 3000, show_col_types = FALSE, progress = FALSE), error=function(e) NULL)
    return(if (is.null(x)) list() else list(list(sheet="", data=x)))
  }
  if (ext %in% c("tsv","txt")) {
    x <- tryCatch(read_tsv(path, n_max = 3000, show_col_types = FALSE, progress = FALSE), error=function(e) NULL)
    if (!is.null(x) && ncol(x) > 1) return(list(list(sheet="", data=x)))
    x <- tryCatch(read_delim(path, delim = NULL, n_max = 3000, show_col_types = FALSE, progress = FALSE), error=function(e) NULL)
    return(if (is.null(x)) list() else list(list(sheet="", data=x)))
  }
  if (ext %in% c("xlsx","xls")) {
    sh <- tryCatch(excel_sheets(path), error=function(e) character())
    out <- list()
    for (s in sh) {
      x <- tryCatch(read_excel(path, sheet=s, n_max=3000), error=function(e) NULL)
      if (!is.null(x)) out[[length(out)+1]] <- list(sheet=s, data=x)
    }
    return(out)
  }
  list()
}

docx_text <- function(path) {
  td <- tempfile("docx_")
  dir.create(td)
  on.exit(unlink(td, recursive=TRUE, force=TRUE), add=TRUE)
  ok <- tryCatch({ unzip(path, exdir=td); TRUE }, error=function(e) FALSE)
  if (!ok) return("")
  xmls <- list.files(file.path(td,"word"), pattern="\\.xml$", full.names=TRUE, recursive=TRUE)
  paste(map_chr(xmls, function(x) {
    tryCatch(xml_text(read_xml(x)), error=function(e) "")
  }), collapse=" ")
}

text_file <- function(path) {
  tryCatch(paste(readLines(path, warn=FALSE, encoding="UTF-8"), collapse=" "),
           error=function(e) "")
}

files_for_project <- function(project) {
  base <- file.path(DATA_ROOT, project, "00_metadata")
  if (!dir.exists(base)) return(character())
  x <- list.files(base, recursive=TRUE, full.names=TRUE)
  x[file.exists(x) & !dir.exists(x)]
}

domains <- names(patterns)
file_rows <- list()
detected_src <- setNames(lapply(PROJECTS, function(x) setNames(replicate(length(domains), character(), simplify=FALSE), domains)), PROJECTS)
structured_src <- detected_src
unparsed <- setNames(replicate(length(PROJECTS), character(), simplify=FALSE), PROJECTS)

for (project in PROJECTS) {
  fs <- files_for_project(project)
  message("Scanning ", project, ": ", length(fs), " files")

  for (path in fs) {
    ext <- tolower(tools::file_ext(path))
    rel <- sub(paste0("^", gsub("\\\\","/", normalizePath(file.path(DATA_ROOT, project), winslash="/", mustWork=FALSE)), "/?"), "",
               normalizePath(path, winslash="/", mustWork=FALSE))

    det <- setNames(rep(FALSE, length(domains)), domains)
    struc <- det
    status <- "UNPARSED"
    evidence <- ""

    if (ext %in% c("csv","tsv","txt","xlsx","xls")) {
      tabs <- read_tabular(path)
      if (length(tabs) > 0) status <- "TABULAR_PARSED" else status <- "TABULAR_READ_FAILED"

      ev <- character()
      for (item in tabs) {
        df <- item$data
        if (is.null(df) || ncol(df)==0) next
        hd <- detect_domains(names(df))
        vals <- character()
        for (j in seq_len(min(ncol(df),100))) {
          x <- unique(na.omit(as.character(df[[j]])))
          vals <- c(vals, head(x[x!=""],20))
        }
        vd <- detect_domains(vals)
        for (d in domains) {
          det[d] <- det[d] || hd[d] || vd[d]
          struc[d] <- struc[d] || hd[d]
        }
        ev <- c(ev, paste0(ifelse(item$sheet=="","",paste0("sheet=",item$sheet,"; ")),
                           "columns=", paste(head(names(df),40), collapse=";")))
      }
      evidence <- paste(ev, collapse=" || ")
    }

    if (ext %in% c("md","rmd","r","py","json","xml")) {
      status <- "TEXT_PARSED"
      x <- detect_domains(text_file(path))
      det <- det | x
    }

    if (ext == "docx") {
      status <- "DOCX_TEXT_PARSED"
      x <- detect_domains(docx_text(path))
      det <- det | x
    }

    if (ext %in% c("pdf","rdata","rds","biom","zip","gz","tgz","tar")) {
      status <- "PRESENT_NOT_PARSED"
      unparsed[[project]] <- c(unparsed[[project]], rel)
    }

    fn <- detect_domains(basename(path))
    det <- det | fn

    for (d in domains) {
      if (det[d]) detected_src[[project]][[d]] <- c(detected_src[[project]][[d]], rel)
      if (struc[d]) structured_src[[project]][[d]] <- c(structured_src[[project]][[d]], rel)
    }

    file_rows[[length(file_rows)+1]] <- tibble(
      Project=project,
      RelativePath=rel,
      FileName=basename(path),
      Extension=ext,
      SizeKB=round(file.info(path)$size/1024,2),
      ParseStatus=status,
      DetectedDomains=paste(names(det)[det], collapse=";"),
      StructuredDomains=paste(names(struc)[struc], collapse=";"),
      EvidencePreview=evidence
    )
  }
}

file_inventory <- if (length(file_rows)>0) bind_rows(file_rows) else tibble()

coverage_rows <- list()
manual_rows <- list()
priority_rows <- list()

CORE <- c("patient_id","sample_id","time")
CLINICAL <- c("infection_group","infection_source","outcome","antibiotics","severity")

for (project in PROJECTS) {
  row <- list(Project=project)
  n_struct <- 0
  n_detect <- 0

  for (d in domains) {
    detected <- length(unique(detected_src[[project]][[d]])) > 0
    structured <- length(unique(structured_src[[project]][[d]])) > 0

    st <- if (structured) "STRUCTURED" else if (detected) "PRESENT_UNSTRUCTURED" else "NOT_FOUND"
    if (structured) n_struct <- n_struct + 1
    if (detected) n_detect <- n_detect + 1

    row[[paste0(d,"_status")]] <- st
    row[[paste0(d,"_sources")]] <- paste(head(unique(detected_src[[project]][[d]]),10), collapse=" | ")
  }

  core_status <- vapply(CORE, function(d) row[[paste0(d,"_status")]], character(1))
  ready <- if (all(core_status=="STRUCTURED")) "YES" else if (all(core_status!="NOT_FOUND")) "MANUAL_LINKAGE_REVIEW" else "NO"

  row$CoreMappingReady <- ready
  row$StructuredDomainCount <- n_struct
  row$DetectedDomainCount <- n_detect
  row$UnparsedPotentialFiles <- paste(head(unique(unparsed[[project]]),20), collapse=" | ")

  coverage_rows[[length(coverage_rows)+1]] <- as_tibble(row)

  missing_core <- CORE[vapply(CORE, function(d) row[[paste0(d,"_status")]]=="NOT_FOUND", logical(1))]
  missing_clin <- CLINICAL[vapply(CLINICAL, function(d) row[[paste0(d,"_status")]]=="NOT_FOUND", logical(1))]
  unstruct <- c(CORE,CLINICAL)[vapply(c(CORE,CLINICAL), function(d) row[[paste0(d,"_status")]]=="PRESENT_UNSTRUCTURED", logical(1))]

  if (length(missing_core)>0) {
    pri <- "HIGH"
    act <- "Core patient/sample/time linkage is still missing. Inspect existing supplement/repository first; contact authors only if public data truly lack it."
  } else if (length(unstruct)>0) {
    pri <- "MEDIUM"
    act <- "No new data collection yet. Existing files appear to contain the information but require parsing/linkage."
  } else if (length(missing_clin)>0) {
    pri <- "LOW"
    act <- "Core longitudinal mapping is available. Collect missing clinical covariates only if required by the planned model."
  } else {
    pri <- "NONE"
    act <- "No manual collection currently required."
  }

  if (project=="PRJEB67798" && row$patient_id_status!="STRUCTURED") {
    pri <- "HIGH"
    act <- "Parse existing supplement first. If patient-level linkage is still absent, author contact is likely required because the publication states public source data exclude patient data."
  }

  manual_rows[[length(manual_rows)+1]] <- tibble(
    Project=project,
    Priority=pri,
    MissingCore=paste(missing_core,collapse=";"),
    MissingClinical=paste(missing_clin,collapse=";"),
    PresentButUnstructured=paste(unstruct,collapse=";"),
    RecommendedAction=act
  )

  score <- ifelse(ready=="YES",100,0) + n_struct*10 + n_detect
  next_step <- if (ready=="YES") "BUILD_SAMPLE_PATIENT_TIME_MAP" else if (length(missing_core)==0) "PARSE_EXISTING_RESOURCES" else "FILL_CORE_MAPPING_GAPS"

  priority_rows[[length(priority_rows)+1]] <- tibble(
    Project=project,
    MappingReadiness=ready,
    CoverageScore=score,
    RecommendedNextStep=next_step
  )
}

coverage <- bind_rows(coverage_rows)
manual <- bind_rows(manual_rows)
priority <- bind_rows(priority_rows) |> arrange(desc(CoverageScore), Project)

stamp <- format(Sys.time(), "%Y%m%d_%H%M%S")

p1 <- file.path(OUT_ROOT, paste0("V2_metadata_coverage_audit_",stamp,".csv"))
p2 <- file.path(OUT_ROOT, paste0("V2_metadata_file_inventory_",stamp,".csv"))
p3 <- file.path(OUT_ROOT, paste0("V2_manual_collection_required_",stamp,".csv"))
p4 <- file.path(OUT_ROOT, paste0("V2_mapping_priority_",stamp,".csv"))

write_excel_csv(coverage,p1,na="")
write_excel_csv(file_inventory,p2,na="")
write_excel_csv(manual,p3,na="")
write_excel_csv(priority,p4,na="")

cat("\n================ STEP 65 COMPLETE ================\n")
cat("R version:", R.version.string, "\n\n")
print(manual |> select(Project,Priority,MissingCore,MissingClinical,PresentButUnstructured), n=Inf, width=Inf)
cat("\nOutputs:\n",p1,"\n",p2,"\n",p3,"\n",p4,"\n")
cat("==================================================\n")

File: analysis/65_audit_v2_metadata_coverage.py
# -*- coding: utf-8 -*-
from __future__ import annotations

import re
import zipfile
import xml.etree.ElementTree as ET
from pathlib import Path
from datetime import datetime

import pandas as pd

DATA_ROOT = Path(r"E:\sepsis_project\data")
OUT_ROOT = Path(r"E:\sepsis_project\results\V2_05_metadata_coverage")
OUT_ROOT.mkdir(parents=True, exist_ok=True)

TARGET_PROJECTS = [
    "PRJEB37289",
    "PRJEB67798",
    "PRJEB82425",
    "PRJNA516701",
    "PRJNA578267",
    "PRJNA595346",
    "PRJNA884103",
]

DOMAINS = {
    "patient_id": [r"\bpatient\b", r"\bsubject\b", r"participant", r"patient.?id", r"subject.?id", r"individual.?id"],
    "sample_id": [r"sample.?id", r"sample_accession", r"biosample", r"sequence.?name", r"sample.?name", r"specimen"],
    "run_id": [r"run_accession", r"sra.?run", r"\brun.?id\b", r"\bcrr\d+", r"\bsrr\d+", r"\berr\d+"],
    "time": [r"time.?point", r"timepoint", r"\bday\b", r"\bdol\b", r"collection.?date", r"sampling.?date", r"visit", r"baseline", r"discharge", r"intubation", r"post.?antibiotic", r"post.?admission"],
    "infection_group": [r"\bsepsis\b", r"septic.?shock", r"\bvap\b", r"pneumonia", r"infected", r"infection.?group", r"bacteremia", r"bacteraemia", r"bloodstream.?infection"],
    "infection_source": [r"infection.?source", r"infection.?site", r"site.?of.?infection", r"abdominal.?origin", r"pulmonary.?infection", r"urinary.?infection", r"bloodstream.?infection"],
    "outcome": [r"\boutcome\b", r"mortality", r"\bdeath\b", r"\bdied\b", r"survival", r"survivor", r"deceased", r"icu.?length.?of.?stay", r"hospital.?length.?of.?stay", r"ventilator.?free", r"adverse.?outcome"],
    "antibiotics": [r"antibiotic", r"antimicrobial", r"\babx\b", r"cefazolin", r"clindamycin", r"tobramycin", r"colistin", r"cefotaxime", r"antibiotic.?score"],
    "severity": [r"\bsofa\b", r"\bapache\b", r"severity", r"shock", r"vasopressor", r"mechanical.?ventilation", r"ventilation.?duration"],
    "demographics": [r"\bage\b", r"\bsex\b", r"gender", r"birth.?weight", r"gestational.?age", r"\bbmi\b"],
    "body_site": [r"body.?site", r"sample.?type", r"rectal", r"stool", r"faecal", r"fecal", r"tracheal", r"endotracheal", r"gut", r"lung"],
}

TABULAR_EXTS = {".csv",".tsv",".txt",".xlsx",".xls"}
TEXT_EXTS = {".md",".rmd",".r",".py",".json",".xml"}
UNPARSED_EXTS = {".pdf",".rdata",".rds",".biom",".zip",".gz"}
MAX_TEXT_CHARS = 2_000_000
MAX_TABULAR_ROWS = 3000

def detect_domains(text: str):
    t = str(text).lower()
    out = {}
    for domain, pats in DOMAINS.items():
        out[domain] = any(re.search(p, t, re.I) for p in pats)
    return out

def read_tabular(path: Path):
    ext = path.suffix.lower()

    if ext == ".csv":
        try:
            df = pd.read_csv(path, dtype=str, nrows=MAX_TABULAR_ROWS, low_memory=False, encoding_errors="replace")
            return ({"sheet":"","df":df},)
        except Exception:
            return tuple()

    if ext in {".tsv",".txt"}:
        try:
            df = pd.read_csv(path, sep="\t", dtype=str, nrows=MAX_TABULAR_ROWS, low_memory=False, encoding_errors="replace")
            if df.shape[1] > 1:
                return ({"sheet":"","df":df},)
        except Exception:
            pass
        try:
            df = pd.read_csv(path, sep=None, engine="python", dtype=str, nrows=MAX_TABULAR_ROWS, on_bad_lines="skip", encoding_errors="replace")
            return ({"sheet":"","df":df},)
        except Exception:
            return tuple()

    if ext in {".xlsx",".xls"}:
        try:
            xl = pd.ExcelFile(path)
        except Exception:
            return tuple()
        out = []
        for sheet in xl.sheet_names:
            try:
                df = pd.read_excel(path, sheet_name=sheet, dtype=str, nrows=MAX_TABULAR_ROWS)
                out.append({"sheet":sheet,"df":df})
            except Exception:
                pass
        return tuple(out)

    return tuple()

def docx_text(path: Path):
    try:
        with zipfile.ZipFile(path) as z:
            parts = []
            for member in z.namelist():
                if member.startswith("word/") and member.endswith(".xml"):
                    try:
                        root = ET.fromstring(z.read(member))
                        parts.extend(x.text for x in root.iter() if x.text)
                    except Exception:
                        pass
            return " ".join(parts)[:MAX_TEXT_CHARS]
    except Exception:
        return ""

def text_file(path: Path):
    try:
        return path.read_text(encoding="utf-8", errors="replace")[:MAX_TEXT_CHARS]
    except Exception:
        return ""

def project_files(project: str):
    base = DATA_ROOT / project / "00_metadata"
    if not base.exists():
        return []
    return [p for p in base.rglob("*") if p.is_file()]

def main():
    stamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    file_rows = []
    project_domain_sources = {p:{d:[] for d in DOMAINS} for p in TARGET_PROJECTS}
    project_structured_sources = {p:{d:[] for d in DOMAINS} for p in TARGET_PROJECTS}
    project_unparsed = {p:[] for p in TARGET_PROJECTS}

    for project in TARGET_PROJECTS:
        files = project_files(project)
        print(f"Scanning {project}: {len(files)} files")

        for path in files:
            ext = path.suffix.lower()
            rel = str(path.relative_to(DATA_ROOT / project))
            detected = {d:False for d in DOMAINS}
            structured = {d:False for d in DOMAINS}
            evidence = []
            parse_status = "UNPARSED"

            if ext in TABULAR_EXTS:
                tables = read_tabular(path)
                if tables:
                    parse_status = "TABULAR_PARSED"

                for item in tables:
                    df = item["df"]
                    cols = [str(c) for c in df.columns]
                    hdet = detect_domains(" | ".join(cols))

                    vals = []
                    try:
                        for c in df.columns[:100]:
                            vals.extend(
                                df[c].dropna().astype(str).drop_duplicates().head(20).tolist()
                            )
                    except Exception:
                        pass
                    vdet = detect_domains(" | ".join(vals)[:MAX_TEXT_CHARS])

                    for d in DOMAINS:
                        if hdet[d] or vdet[d]:
                            detected[d] = True
                        if hdet[d]:
                            structured[d] = True

                    prefix = f"sheet={item['sheet']}; " if item["sheet"] else ""
                    evidence.append(prefix + "columns=" + ";".join(cols[:40]))

            elif ext in TEXT_EXTS:
                parse_status = "TEXT_PARSED"
                ddet = detect_domains(text_file(path))
                for d in DOMAINS:
                    detected[d] = ddet[d]

            elif ext == ".docx":
                parse_status = "DOCX_TEXT_PARSED"
                ddet = detect_domains(docx_text(path))
                for d in DOMAINS:
                    detected[d] = ddet[d]

            elif ext in UNPARSED_EXTS:
                project_unparsed[project].append(rel)
                parse_status = "PRESENT_NOT_PARSED"

            filename_det = detect_domains(path.name)
            for d in DOMAINS:
                if filename_det[d]:
                    detected[d] = True
                if detected[d]:
                    project_domain_sources[project][d].append(rel)
                if structured[d]:
                    project_structured_sources[project][d].append(rel)

            file_rows.append({
                "Project": project,
                "RelativePath": rel,
                "FileName": path.name,
                "Extension": ext,
                "SizeKB": round(path.stat().st_size/1024,2),
                "ParseStatus": parse_status,
                "DetectedDomains": ";".join(d for d,v in detected.items() if v),
                "StructuredDomains": ";".join(d for d,v in structured.items() if v),
                "EvidencePreview": " || ".join(evidence)[:3000],
            })

    coverage_rows = []
    manual_rows = []
    priority_rows = []
    CORE = ["patient_id","sample_id","time"]
    CLINICAL = ["infection_group","infection_source","outcome","antibiotics","severity"]

    for project in TARGET_PROJECTS:
        row = {"Project":project}
        structured_count = 0
        detected_count = 0

        for d in DOMAINS:
            detected = bool(project_domain_sources[project][d])
            structured = bool(project_structured_sources[project][d])
            if structured:
                status = "STRUCTURED"
                structured_count += 1
            elif detected:
                status = "PRESENT_UNSTRUCTURED"
            else:
                status = "NOT_FOUND"
            if detected:
                detected_count += 1
            row[f"{d}_status"] = status
            row[f"{d}_sources"] = " | ".join(list(dict.fromkeys(project_domain_sources[project][d]))[:10])

        core_structured = all(row[f"{d}_status"]=="STRUCTURED" for d in CORE)
        if core_structured:
            mapping_ready = "YES"
        elif all(row[f"{d}_status"]!="NOT_FOUND" for d in CORE):
            mapping_ready = "MANUAL_LINKAGE_REVIEW"
        else:
            mapping_ready = "NO"

        row["CoreMappingReady"] = mapping_ready
        row["StructuredDomainCount"] = structured_count
        row["DetectedDomainCount"] = detected_count
        row["UnparsedPotentialFiles"] = " | ".join(project_unparsed[project][:20])
        coverage_rows.append(row)

        missing_core = [d for d in CORE if row[f"{d}_status"]=="NOT_FOUND"]
        missing_clin = [d for d in CLINICAL if row[f"{d}_status"]=="NOT_FOUND"]
        unstructured = [d for d in CORE+CLINICAL if row[f"{d}_status"]=="PRESENT_UNSTRUCTURED"]

        if missing_core:
            level = "HIGH"
            action = "MANUAL_COLLECTION_REQUIRED: obtain patient/sample/time linkage from supplement, repository, or authors."
        elif unstructured:
            level = "MEDIUM"
            action = "NO NEW DATA COLLECTION YET: parse/link existing supplement or repository resources first."
        elif missing_clin:
            level = "LOW"
            action = "CORE MAPPING AVAILABLE; collect missing clinical covariates only if required by the planned model."
        else:
            level = "NONE"
            action = "NO MANUAL COLLECTION CURRENTLY REQUIRED."

        if project == "PRJEB67798" and row["patient_id_status"]!="STRUCTURED":
            level = "HIGH"
            action = "AFTER PARSING EXISTING SUPPLEMENT: if patient-level linkage is absent, author contact is likely required because public source data exclude patient data."

        manual_rows.append({
            "Project":project,
            "Priority":level,
            "MissingCore":";".join(missing_core),
            "MissingClinical":";".join(missing_clin),
            "PresentButUnstructured":";".join(unstructured),
            "RecommendedAction":action,
        })

        score = (100 if core_structured else 0) + structured_count*10 + detected_count
        priority_rows.append({
            "Project":project,
            "MappingReadiness":mapping_ready,
            "CoverageScore":score,
            "RecommendedNextStep":(
                "BUILD_SAMPLE_PATIENT_TIME_MAP" if core_structured
                else "PARSE_EXISTING_RESOURCES" if not missing_core
                else "FILL_CORE_MAPPING_GAPS"
            ),
        })

    coverage = pd.DataFrame(coverage_rows)
    files = pd.DataFrame(file_rows)
    manual = pd.DataFrame(manual_rows)
    priority = pd.DataFrame(priority_rows).sort_values(["CoverageScore","Project"], ascending=[False,True])

    p1 = OUT_ROOT / f"V2_metadata_coverage_audit_{stamp}.csv"
    p2 = OUT_ROOT / f"V2_metadata_file_inventory_{stamp}.csv"
    p3 = OUT_ROOT / f"V2_manual_collection_required_{stamp}.csv"
    p4 = OUT_ROOT / f"V2_mapping_priority_{stamp}.csv"

    coverage.to_csv(p1,index=False,encoding="utf-8-sig")
    files.to_csv(p2,index=False,encoding="utf-8-sig")
    manual.to_csv(p3,index=False,encoding="utf-8-sig")
    priority.to_csv(p4,index=False,encoding="utf-8-sig")

    print()
    print("================ STEP 05 COMPLETE ================")
    print(manual[["Project","Priority","MissingCore","MissingClinical","RecommendedAction"]].to_string(index=False))
    print()
    print(p1)
    print(p2)
    print(p3)
    print(p4)

if __name__ == "__main__":
    main()
